### PR TITLE
ARROW-1785: [Format/C++/Java] Remove VectorLayout from serialized schemas

### PR DESCRIPTION
**Pull request by wesm**
What's here so far removes this code from the Flatbuffers schema and the C++ implementation. This logic is a little bit entangled on the Java side, I need some help from @icexelloss or @BryanCutler or @julienledem or someone else to handle the Java refactoring. It might be easiest to preserve the ArrowVectorType/TypeLayout/VectorLayout objects for now but simply remove the Flatbuffers dependency (we do need the names of the vectors in the JSON files used for integration testing)

cc @trxcllnt since we may want to roll in the JS changes in this patch

### CODE REVIEW

**Comment by trxcllnt** [comment]
@wesm I'll regenerate the JS flatbuffers locally, and add to #1294. I have a bunch of little questions about the layouts for each type, but I'll ping you over email/slack to get the details.

**Comment by icexelloss** [comment]
I can take a look at this. My understanding is we want to remove vector layout from metadata and assume buffer ordering (validity, data) and (validity, offset, data) in readers, is that correct?

**Comment by wesm** [comment]
Right. Unfortunately the Java implementation (including the JSON reader/writer) is a bit intertwined with the Flatbuffers

**Comment by icexelloss** [comment]
Gee, also the concurrent patches that touch reader/writer is getting a bit out of control.

https://github.com/apache/arrow/pull/1290
https://github.com/apache/arrow/pull/1259

And this one. cc @siddharthteotia @BryanCutler.

**Comment by wesm** [comment]
I think here I would like to touch the bare minimum of Java code to remove the dependence on the Flatbuffers. I can try to do this today so not impact the java-vector-refactor branch

**Comment by wesm** [comment]
This is a fairly invasive refactor. I don't have the skills or time to do the Java work, someone else is going to have to help. Having the Java object model tightly coupled to the Flatbuffers schemas is not great. I hope we can fix that

**Comment by icexelloss** [comment]
I can take a look at Java side next week.

**Comment by icexelloss** [comment]
@wesm I took a look today. The change does't seem to be too bad. However, json reader probably needs the new vector class hierarchy (BaseNullableFixedWidthVector, BaseNullableVariableWidthVector) to determine the expected layout.

So it's probably best either wait until java-refactor gets merged to master or do it on java-refactor branch. cc @siddharthteotia

**Comment by wesm** [comment]
Cool, since the merge seems likely this week, we can just wait. thanks!

**Comment by wesm** [comment]
I'll rebase this, then I think we can do the Java work and merge this

**Comment by wesm** [comment]
Rebased

**Comment by wesm** [comment]
Can someone look at this early this coming week? We should not release 0.8.0 without this

**Comment by icexelloss** [comment]
@wesm, I am at pydata today but I will try to take a look later today or tomorrow.

**Comment by wesm** [comment]
@siddharthteotia can you review this? This is the last format-related change required for 0.8.0

**Comment by siddharthteotia** [comment]
Looks like the java side of changes are messed up because of rebase commits? I am seeing code changes (like removal of non-nullable vectors) and those are not the java changes made by this PR. It is hard to make out what are we doing on the JAVA side as far this PR is concerned.

I can see that ArrowVectorType  and VectorLayout have been removed and we are using BufferType and BufferLayout instead. Is that correct? 

I believe that these objects are still part of serialized schema -- it's just that names have changed.  Instead of ArrowVectorType, we have BufferType and instead of VectorLayout we have BufferLayout. Is the correct summary of JAVA side changes?

**Comment by icexelloss** [comment]
@siddharthteotia `VectorLayout` and `TypeLayout` are no longer part of the serialized schema. 

`VectorLayout` and `ArrowVectorType` are renamed to `BufferLayout` and `BufferType` and they are just internal classes now. 

Other changes to Java are to Reader/Writer classes 0- they no longer read/write type layout when reading/writing schema.

**Comment by wesm** [comment]
Rebased, apologies for the diff noise

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java [hunk 1]
```diff
@@ -454,7 +454,7 @@ public void testNestedListVector() throws Exception {
       /* allocate memory */
       listWriter.allocate();
 
-      /* the dataVector that backs a listVector will also be a
+      /* the dataBuffer that backs a listVector will also be a
```
Comment (wesm) [inline java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: Maybe some of these code comment changes aren't quite right?

**Review by wesm — COMMENTED** [review]
Minor comment. From an Java API perspective these changes appear non-disruptive

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java (hunk 1 above)
Comment (icexelloss) [reply java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: These are likely done be the IDE, I will double check.

Comment (icexelloss) [reply java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: Fixed here https://github.com/wesm/arrow/pull/3

**Comment by siddharthteotia** [comment]
+1

**Comment by wesm** [comment]
Thanks all!!

**Pull request closed without merging**